dynamic_programming/dp_bitmasking.cpp

```diff
@@ -0,0 +1,132 @@
+/**
+ * @file
+ * @brief Implementation of Bitmasking DP approach
+ * (https://codeforces.com/blog/entry/337)
+ *
+ * @details
+ * Given Integer N and Array of size 2*N. Make N pairs using each element
+ * exactly once such that following score is maximized.
+ * score = summation(gcd(x_i, y_i) * i)
+ * for 1 <= i <= N, (x_i, y_i) is i-th pair.
+ * N <= 10
+ * 1 <= A[i] <= 1e9
+ *
+ * ### Algorithm
+ * For each Mask 0 to 111....(2*N times) Add new pair to this set from yet
+ * unselected elements and try to maximize the newly formed mask after
+ * adding these elements with value formed after adding new pair.
+ * the ans we are looking for will be in mask which shows all elements are
+ * added, that is 111...(2*N times).
+ *
+ * @author [Syed Faizan](https://github.com/faizan2700)
+ */
+#include <algorithm>  /// for calculating gcd
+#include <array>      /// for creating array
+#include <cassert>    /// for using assert in self-test
+#include <iostream>   /// for IO operations
+#include <vector>     /// for std::vector
+
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
+namespace dynamic_programming {
+/**
+ * @namespace dp_bitmask
+ * @brief Implementation of DP with Bitmasking
+ */
+namespace dp_bitmask {
+/**
+ * @brief Find the maximum value of a new mask by adding
+ * a pair to a previously formed mask.
+ * @tparam T size of the array A.
+ * @param A an array of elements from which pairs will be made.
+ * @param N size of an array.
+ * @return maximum score which can be formed with pairs of this array.
+ */
+template <size_t T>
+uint64_t find_max_score(std::array<uint32_t, T> A, const uint8_t N) {

```

Comment:
Better to pass this as const ref.
```suggestion
uint64_t find_max_score(const std::array<uint32_t, T> &A, const uint8_t N) {

```

Reply:
ok

Reply:
Thanks, it's just a bit of performance improvement.